File: 11. Insert into BST.cpp
#/*
 Task: Insert into BST
 Time Complexity: O(h) — height of tree (O(log n) for balanced BST)
 Space Complexity: O(h) — recursion stack
*/
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
Node* insertBST(Node* root, int key)
{
    if (!root)
        return new Node(key);
    if (key < root->data)
        root->left = insertBST(root->left, key);
    else
        root->right = insertBST(root->right, key);
    return root;
}
void inorder(Node* root)
{
    if (!root)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}
int main()
{
    Node* root = nullptr;
    root = insertBST(root, 5);
    insertBST(root, 3);
    insertBST(root, 7);
    insertBST(root, 2);
    insertBST(root, 4);
    cout << "BST inorder: ";
    inorder(root);
    cout << endl;
    return 0;
}
// sc tc

File: 11. Insert into BST.py
"""
Task: Insert into BST
Time Complexity: O(h)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def insert_bst(root,key):
    if not root: return Node(key)
    if key<root.data:
        root.left=insert_bst(root.left,key)
    else:
        root.right=insert_bst(root.right,key)
    return root

def inorder(root):
    if not root: return
    inorder(root.left); print(root.data, end=' '); inorder(root.right)

if __name__=='__main__':
    root=None
    root=insert_bst(root,5)
    insert_bst(root,3); insert_bst(root,7); insert_bst(root,2); insert_bst(root,4)
    print('BST inorder:', end=' '); inorder(root); print()
